summingbird-storm/src/main/scala/com/twitter/summingbird/storm/builder/Topology.scala

```diff
@@ -0,0 +1,156 @@
+  }
+
+  def withBolt[I, O](id: String, bolt: Topology.Bolt[I, O]): (Topology.BoltId[I, O], Topology) = {
+    val boltId = Topology.BoltId[I, O](id)
+    assert(!bolts.contains(boltId))
+    (boltId, Topology(spouts, bolts.updated(boltId, bolt), edges))
+  }
+
+  def withEdge[T](edge: Topology.Edge[T]): Topology = {
+    assert(contains(edge.source) && contains(edge.dest))
+    assert(edge.source != edge.dest)
+    assert(edges.forall { anotherEdge =>
+      !(anotherEdge.source == edge.source && anotherEdge.dest == edge.dest)
+    })
+
+    Topology(spouts, bolts, edges ++ List(edge))
+  }
+
+  def contains(id: Topology.ComponentId): Boolean = id match {
+    case spoutId: Topology.SpoutId[_] => spouts.contains(spoutId)
+    case boltId: Topology.BoltId[_, _] => bolts.contains(boltId)
+  }
+
+  def incomingEdges[T](id: Topology.ReceivingId[T]): List[Topology.Edge[T]] = {
+    assert(contains(id))
+    edges.filter(_.dest == id).asInstanceOf[List[Topology.Edge[T]]]
+  }
+
+  def outcomingEdges[T](id: Topology.EmittingId[T]): List[Topology.Edge[T]] = {
+    assert(contains(id))
+    edges.filter(_.source == id).asInstanceOf[List[Topology.Edge[T]]]
+  }
+
+  def build(jobId: JobId): StormTopology = {
+    val builder = new TopologyBuilder
+
+    spouts.foreach { case (spoutId: Topology.SpoutId[Any], spout: Topology.Spout[Any]) =>
+      val builtSpout = SpoutBuilder[Any](
+        jobId,
+        spoutId,
+        spout,
+        outcomingEdges(spoutId)
+      )
+      builder.setSpout(spoutId.id, builtSpout, spout.parallelism)
+    }
+
+    bolts.foreach { case (boltId: Topology.BoltId[Any, Any], bolt: Topology.Bolt[Any, Any]) =>
+      val builtBolt: IRichBolt = BoltBuilder.apply[Any, Any](
+        jobId,
+        boltId,
+        bolt,
+        incomingEdges(boltId),
+        outcomingEdges(boltId)
+      )
+
+      val declarer = builder.setBolt(
+        boltId.id,
+        builtBolt,
+        bolt.parallelism
+      ).addConfigurations(tickConfig)
+
+      incomingEdges(boltId).foreach { edge =>
+        edge.edgeType.grouping.apply(declarer, edge.source)
+      }
+    }
+
+    builder.createTopology()
+  }
+
+  /**
+   * Set storm to tick our nodes every second to clean up finished futures
+   */
+  private def tickConfig = {
+    val boltConfig = new BacktypeStormConfig
+    boltConfig.put(BacktypeStormConfig.TOPOLOGY_TICK_TUPLE_FREQ_SECS, java.lang.Integer.valueOf(1))
+    boltConfig
+  }
+}
+
+object Topology {
+  val EMPTY = Topology(Map(), Map(), List())
+
+  sealed trait ComponentId {
+    val id: String
+  }
+
+  trait ReceivingId[-I] extends ComponentId
+
+  trait EmittingId[+O] extends ComponentId
+
+  case class SpoutId[+O](override val id: String) extends EmittingId[O]
+
+  case class BoltId[-I, +O](override val id: String) extends ReceivingId[I] with EmittingId[O]
+
+  case class Edge[T](source: EmittingId[T], edgeType: EdgeType[T], dest: ReceivingId[T])
+
+  sealed trait Component {
```

Comment:
I shrinked access for all new classes - to `private[summingbird]` for things which used outside and to `private[builder]` for all other things.

I added comments to most places in `Topology.scala` but not for everything, do you think we should add extensive comments? I'm not sure how stable it is, so I prefer do not add long comments for now.